man/dmsystem.1
.TH "DMSYSTEM" "1" "" "Version 2.0.0" "User Commands"
.SH NAME
dmsystem \(en monitors system parameters
.SH SYNOPSIS
\f[B]dmsystem\f[R] \-\-help
.PP
\f[B]dmsystem\f[R] \-\-version
.PP
\f[B]dmsystem\f[R] \-\-\f[B]node\f[R] \f[I]id\f[R] \-\-\f[B]sensor\f[R]
\f[I]id\f[R] \-\-\f[B]target\f[R] \f[I]id\f[R] [\-\-\f[B]receiver\f[R]
\f[I]name\f[R]] [\-\-\f[B]interval\f[R] \f[I]seconds\f[R]]
[\-\-\f[B]count\f[R] \f[I]n\f[R]] [\-\-\f[B]name\f[R] \f[I]name\f[R]]
[\-\-\f[B]logger\f[R] \f[I]name\f[R]] [\-\-\f[B]debug\f[R]]
[\-\-\f[B]verbose\f[R]]
.PP
\f[B]dmsystem\f[R] \-\-\f[B]config\f[R] \f[I]file\f[R]
[\-\-\f[B]name\f[R] \f[I]name\f[R]]
.SH DESCRIPTION
A system monitor to watch free disk space and disk capacity, database
sizes, load average, CPU temperature, and system uptime.
The file system for the free disk space and disk capacity responses is
determined through a given file or directory path.
The paths to the log and observation databases have to be given in the
configuration file.
.TP
cpu_temp
CPU temperature [°C].
Disabled by default.
.TP
disk_free
Free disk space [Byte].
Requires file or directory path.
.TP
disk_capacity
Disk capacity [%].
Requires file or directory path.
.TP
load_avg1
Load average, last minute.
Enabled by default.
.TP
load_avg5
Load average, last 5 minutes.
Enabled by default.
.TP
load_avg15
Load average, last 15 minutes.
Enabled by default.
.TP
uptime
System uptime [sec].
Enabled by default.
.SH OPTIONS
.TP
\f[B]\-\-config\f[R], \f[B]\-c\f[R] \f[I]file\f[R]
File path to the configuration file.
.TP
\f[B]\-\-count\f[R], \f[B]\-C\f[R] \f[I]n\f[R]
Number of observations to create.
No limit if count is set to 0 (default).
.TP
\f[B]\-\-debug\f[R], \f[B]\-D\f[R]
Forward logs messages of level \f[CR]LL_DEBUG\f[R] via IPC (if logger is
set).
.TP
\f[B]\-\-help\f[R], \f[B]\-h\f[R]
Print available command\-line arguments and quit.
.TP
\f[B]\-\-interval\f[R], \f[B]\-I\f[R] \f[I]seconds\f[R]
Interval in seconds in which to read the system parameters.
.TP
\f[B]\-\-logger\f[R], \f[B]\-l\f[R] \f[I]name\f[R]
Name of logger.
If set, sends logs to \f[I]dmlogger(1)\f[R] process of given name.
.TP
\f[B]\-\-name\f[R], \f[B]\-n\f[R] \f[I]name\f[R]
Name of program instance and configuration (default is
\f[CR]dmsystem\f[R]).
.TP
\f[B]\-\-node\f[R], \f[B]\-N\f[R] \f[I]id\f[R]
Node id.
.TP
\f[B]\-\-receiver\f[R], \f[B]\-r\f[R] \f[I]name\f[R]
Name of the observation receiver, without leading \f[CR]/\f[R].
.TP
\f[B]\-\-sensor\f[R], \f[B]\-S\f[R] \f[I]id\f[R]
Sensor id.
.TP
\f[B]\-\-target\f[R], \f[B]\-T\f[R] \f[I]id\f[R]
Target id.
.TP
\f[B]\-\-verbose\f[R], \f[B]\-V\f[R]
Print log messages to \f[I]stderr\f[R].
.TP
\f[B]\-\-version\f[R], \f[B]\-v\f[R]
Print version information and quit.
.SH EXIT STATUS
.TP
\f[B]0\f[R]
Success.
Process terminated without errors.
.TP
\f[B]1\f[R]
Failure.
Process failed.
.SH ENVIRONMENT
.TP
\f[B]DM_LOGGER\f[R]
Name of logger instance to send logs to.
.TP
\f[B]NO_COLOR\f[R]
Disable ANSI colour output.
.SH EXAMPLE
Read system parameters in an interval of 60 seconds and forward
observations to \f[CR]dmrecv\f[R]:
.IP
.EX
$ dmsystem \-N dummy\-node \-S dummy\-sensor \-T dummy\-target \(rs
  \-I 60 \-r dmrecv \-V
.EE
.PP
Start \f[CR]dmrecv\f[R] with:
.IP
.EX
$ dmrecv \-n dmrecv \-t observ \-f jsonl \-o \(dq\-\(dq \-V
.EE


CHANGELOG.md
# Change Log

All notable changes to the project will be documented in this file.

## [2.0.0] – Unreleased

## Library

* `Added` error codes
* `Added` _fortran-nng_ interface bindings to NNG to `vendor/`
* `Added` module `dm_arg_parser` (outsourced from `dm_arg`)
* `Added` module `dm_group` for observation groups
* `Added` module `dm_ipc` for socket access
* `Added` module `dm_ipc_async` for asynchronous I/O
* `Added` module `dm_ipc_disco` for service discovery
* `Added` module `dm_ipc_message` for message handling
* `Added` module `dm_ipc_mutex` for mutex access
* `Added` module `dm_ipc_thread` for thread access
* `Added` module `dm_ipc_type` of IPC enumerators
* `Added` module `dm_job_list` (outsourced from `dm_job`)
* `Changed` names of modules `dm_fifo`, `dm_mqueue`, `dm_mqueue_util`, `dm_pipe`,
  `dm_sem`, `dm_signal`, `dm_thread`, `dm_tty` to include `posix` prefix
* `Changed` name of module `dm_system` to `dm_posix`
* `Changed` database schema due to new data models
* `Changed` default database suffix from `.sqlite` to `.db`
* `Changed` GeoCOM API for Lua to accept prototype observation as first argument
* `Changed` model of observation data structure `dm_observ`
* `Deleted` model of request data strucutre `dm_request`

## Programs

* `Changed` response value units in `dmved`

## Documentation

* Converted user guide from AsciiDoc to Markdown
* Converted man pages from AsciiDoc to Markdown

## [1.0.0] – Unreleased

## Library

* `Added` Bessel, Butterworth, Chebyshev filters to `dm_filter`

## [0.9.9] – 2026-01-11

### Library

* `Added` module `dm_coord` for coordinate transformation
* `Added` module `dm_filter` for filtering
* `Added` module `dm_ghostscript` for Ghostscript access
* `Added` module `dm_netstring` for netstring parsing
* `Added` module `dm_statistics` of statistics functions
* `Added` file tree size function to module `dm_file`
* `Added` file modification date/time to function `dm_file_touch()` in `dm_file`
* `Added` format `FORMAT_TSV` for tab-separated values to `dm_format`
* `Added` swap routines to `dm_util`
* `Changed` error code order for future additions
* `Changed` default database cache size of `dm_db_open()` to improve performance

### Programs

* `Added` author option to `dmreport`
* `Added` meta data to PDF output of `dmreport`
* `Fixed` reading of options in `dmdbctl`
* `Changed` last modified date/time to last log date/time in output of `dmfeed`

## [0.9.8] – 2025-08-30

### Library

* `Added` module `dm_image` for image handling
* `Added` module `dm_js` for JavaScript generation
* `Added` module `dm_roff` for GNU roff abstraction
* `Added` module `dm_serial` for derived type serialisation
* `Added` module `dm_transfer` for HTTP-RPC API file transfers
* `Added` GeoJSON Feature Collection procedures to `dm_geojson`
* `Added` image upload to `dm_rpc`
* `Added` HTTP request and response headers to `dm_rpc`
* `Added` HTTP response headers to `dm_fcgi`
* `Added` terminals `gpic`, `postscript`, and `sixeltek` to `dm_plot`
* `Added` image and transfer functions to `dm_db_api`
* `Changed` structure of database abstraction layer, added modules `dm_db_api`, `dm_db_count`, `dm_db_json`, `dm_db_pragma`, and `dm_db_row`
* `Changed` API of modules `dm_arg` and `dm_config` to object-oriented
* `Changed` query parameter API in `dm_cgi`
* `Changed` Git submodules to copies of third-party dependencies
* `Fixed` reading from pipe in `dm_pipe`
* `Fixed` unsigned type conversion in `dm_c`
* `Fixed` return code handling of database select functions

### Programs

* `Added` program `dmcamera` for capturing of camera images
* `Added` program `dmupload` for image upload to HTTP-RPC API
* `Added` scale factor for response values to `dmreport`
* `Added` PDF and PostScript output to `dmreport`
* `Added` message queue reading to `dmmb`
* `Added` image upload to `dmapi`
* `Added` image viewer to `dmweb`
* `Added` environment variables `DM_IMAGE_DB`, `DM_IMAGE_DIR` to `dmapi` and `dmweb`
* `Added` log file output to `dmlogger`
* `Changed` environment variables in `dmapi` and `dmweb` to `DM_BEAT_DB`, `DM_LOG_DB`, `DM_OBSERV_DB`
* `Changed` derived type serialisation in `dmapi` and `dmexport`
* `Changed` database access in `dmexport`
* `Changed` configuration of `dmgrc`

## [0.9.7] – 2025-04-05

### Library

* `Added` module `dm_c` for C interoperability
* `Added` module `dm_camera` for webcam and RTSP access (FFmpeg)
* `Added` module `dm_crypto` of cryptographic hash functions
* `Added` module `dm_db_query` for SQL query building
* `Added` module `dm_dwd` for DWD API format parsing
* `Added` module `dm_dwd_api` for DWD API access (libcurl)
* `Added` module `dm_ftp` for FTP transfer (libcurl)

app/dmsystem.f90
! dmsystem.f90
!
! Author:  Philipp Engel
! Licence: ISC
program dmsystem
    !! System monitor. Sends system status as observation to receiver via POSIX
    !! message queue.
    use :: dmpack
    implicit none (type, external)

    character(*), parameter :: APP_NAME  = 'dmsystem'
    integer,      parameter :: APP_MAJOR = 2
    integer,      parameter :: APP_MINOR = 0
    integer,      parameter :: APP_PATCH = 0

    character(*), parameter :: APP_OBSERV_NAME  = 'system_status' !! Name of all observations.
    character(*), parameter :: APP_REQUEST_NAME = 'status'        !! Name of all observation requests.

    logical, parameter :: APP_MQ_BLOCKING = .true. !! Observation forwarding is blocking.

    type :: app_options_type
        !! Enabled responses.
        character(FILE_PATH_LEN) :: disk_free  = ' '     !! Disk free path (file or directory).
        character(FILE_PATH_LEN) :: log_db     = ' '     !! Path of log database.
        character(FILE_PATH_LEN) :: observ_db  = ' '     !! Path of log database.
        logical                  :: cpu_temp   = .false. !! Enable CPU temperature.
        logical                  :: load_avg1  = .true.  !! Enable system load average, 1 min.
        logical                  :: load_avg5  = .true.  !! Enable system load average, 5 min.
        logical                  :: load_avg15 = .true.  !! Enable system load average, 15 min.
        logical                  :: uptime     = .true.  !! Enable system uptime.
    end type app_options_type

    type :: app_type
        !! Application settings.
        character(ID_LEN)              :: name      = APP_NAME !! Name of instance/configuration.
        character(FILE_PATH_LEN)       :: config    = ' '      !! Path to configuration file.
        character(LOGGER_NAME_LEN)     :: logger    = ' '      !! Name of logger (name implies IPC).
        character(NODE_ID_LEN)         :: node_id   = ' '      !! Node id (required).
        character(SENSOR_ID_LEN)       :: sensor_id = ' '      !! Sensor id (required).
        character(TARGET_ID_LEN)       :: target_id = ' '      !! Target id (required).
        character(OBSERV_RECEIVER_LEN) :: receiver  = ' '      !! Observation receiver.
        integer                        :: count     = 0        !! Maximum number of observations to send (0 means unlimited).
        integer                        :: interval  = 600      !! Emit interval in seconds (>= 0).
        logical                        :: debug     = .false.  !! Forward debug messages via IPC.
        logical                        :: verbose   = .false.  !! Print debug messages to stderr.
        type(app_options_type)         :: options              !! Enabled responses.
    end type app_type

    class(logger_class), pointer :: logger ! Logger object.

    integer        :: rc  ! Return code.
    type(app_type) :: app ! App settings.

    call dm_init()

    rc = read_args(app)
    if (dm_is_error(rc)) call dm_stop(STOP_FAILURE)

    logger => dm_logger_get_default()
    call logger%configure(name    = app%logger,  & ! Name of logger process.
                          node_id = app%node_id, & ! Node id.
                          source  = app%name,    & ! Log source.
                          debug   = app%debug,   & ! Forward DEBUG messages via IPC.
                          ipc     = .true.,      & ! Enable IPC (if logger is set).
                          verbose = app%verbose)   ! Print logs to standard error.
    call logger%info('started ' // APP_NAME)

    call dm_posix_signal_register(signal_callback)
    rc = run(app)
    call shutdown(rc)
contains
    integer function run(app) result(rc)
        !! Run system monitoring and emit observations.
        character(*), parameter :: DISABLED = 'disabled'
        character(*), parameter :: ENABLED  = 'enabled'

        type(app_type), intent(inout) :: app !! App type.

        integer :: iter, msec, sec
        logical :: debug

        logical :: has_cpu_temp, has_disk_free, has_log_db, has_observ_db, has_uptime
        logical :: has_load_avg, has_load_avg1, has_load_avg5, has_load_avg15

        type(observ_type) :: observ
        type(timer_type)  :: timer

        debug = (app%debug .or. app%verbose)

        has_cpu_temp   = app%options%cpu_temp
        has_load_avg1  = app%options%load_avg1
        has_load_avg5  = app%options%load_avg5
        has_load_avg15 = app%options%load_avg15
        has_load_avg   = (has_load_avg1 .or. has_load_avg5 .or. has_load_avg15)
        has_uptime     = app%options%uptime

        has_disk_free  = dm_string_has(app%options%disk_free)
        has_log_db     = dm_string_has(app%options%log_db)
        has_observ_db  = dm_string_has(app%options%observ_db)

        if (debug) then
            call logger%debug('running on node ' // trim(app%node_id) // ' as sensor ' // trim(app%sensor_id) // ' with target ' // trim(app%target_id))
            call logger%debug('interation interval set to ' // dm_itoa(app%interval) // ' sec')

            if (app%count > 0) then
                call logger%debug('iteration count set to ' // dm_itoa(app%count))
            else
                call logger%debug('no iteration count set')
            end if

            if (dm_string_has(app%receiver)) then
                call logger%debug('observation forwarding to message queue /' // trim(app%receiver) // ' enabled')
            else
                call logger%debug('observation forwarding disabled')
            end if

            call logger%debug('CPU temperature monitoring '      // dm_btoa(has_cpu_temp,   ENABLED, DISABLED))
            call logger%debug('disk free monitoring '            // dm_btoa(has_disk_free,  ENABLED, DISABLED))
            call logger%debug('load average 1 monitoring '       // dm_btoa(has_load_avg1,  ENABLED, DISABLED))
            call logger%debug('load average 5 monitoring '       // dm_btoa(has_load_avg5,  ENABLED, DISABLED))
            call logger%debug('load average 15 monitoring '      // dm_btoa(has_load_avg15, ENABLED, DISABLED))
            call logger%debug('log database monitoring '         // dm_btoa(has_log_db,     ENABLED, DISABLED))
            call logger%debug('observation database monitoring ' // dm_btoa(has_observ_db,  ENABLED, DISABLED))
            call logger%debug('uptime monitoring '               // dm_btoa(has_uptime,     ENABLED, DISABLED))
        end if

        iter = 1

        emit_loop: do
            call dm_timer_start(timer)
            if (debug .and. app%count > 0) call logger%debug('starting iteration ' // dm_itoa(iter) // '/' // dm_itoa(app%count))

            ! Initialise observation.
            observ = observ_type(id        = dm_uuid4(),      &
                                 node_id   = app%node_id,     &
                                 sensor_id = app%sensor_id,   &
                                 target_id = app%target_id,   &
                                 name      = APP_OBSERV_NAME, &
                                 timestamp = dm_time_now(),   &
                                 source    = app%name)
            rc = dm_observ_add_receiver(observ, app%receiver) ! Ignore empty receiver error.
            if (debug) call logger%debug('created observation ' // observ%name, observ=observ)

            ! Get system parameters.
            if (has_cpu_temp)  call add_cpu_temp    (observ)                                               ! Get CPU temperature.
            if (has_disk_free) call add_disk_free   (observ, app%options%disk_free)                        ! Get available disk space and disk capacity.
            if (has_load_avg)  call add_load_average(observ, has_load_avg1, has_load_avg5, has_load_avg15) ! Get load average.
            if (has_log_db)    call add_log_db      (observ, app%options%log_db)                           ! Get log database size.
            if (has_observ_db) call add_observ_db   (observ, app%options%observ_db)                        ! Get observation database size.
            if (has_uptime)    call add_uptime      (observ)                                               ! Get system uptime.

            ! Forward observation.
            rc = dm_posix_mqueue_forward(observ, name=app%name, blocking=APP_MQ_BLOCKING, use_logger=debug)

            if (debug) call logger%debug('finished observation ' // observ%name, observ=observ)
            if (debug .and. app%count > 0) call logger%debug('finished iteration ' // dm_itoa(iter) // '/' // dm_itoa(app%count))

            if (app%count > 0) then
                iter = iter + 1
                if (iter > app%count) exit emit_loop
            end if

            call dm_timer_stop(timer)
            msec = max(0, int(1000 * (app%interval - dm_timer_result(timer))))
            sec  = dm_msec_to_sec(msec)
            if (debug) call logger%debug('next observation in ' // dm_itoa(sec) // ' sec')
            call dm_msleep(msec)
        end do emit_loop

        if (debug) call logger%debug('finished monitoring')
    end function run

    subroutine shutdown(error)
        !! Stops program.
        integer, intent(in) :: error !! DMPACK error code.

        integer :: stat

        stat = merge(STOP_FAILURE, STOP_SUCCESS, dm_is_error(error))
        call logger%info('stopped ' // APP_NAME, error=error)
        call dm_stop(stat)
    end subroutine shutdown

    ! **************************************************************************
    ! SYSTEM PARAMETER ROUTINES.
    ! **************************************************************************
    subroutine add_cpu_temp(observ)
        !! Adds CPU temperature response to observation.
        character(*), parameter :: NAME = 'cpu_temp'

        type(observ_type), intent(inout) :: observ !! Observation.

        integer :: rc
        real    :: cpu_temp

        rc = dm_posix_cpu_temperature(cpu_temp)

        if (dm_is_error(rc)) then
            call logger%error('failed to read CPU temperature', observ=observ, error=rc)
            return
        end if

        call logger%debug('CPU temperature: ' // dm_ftoa(cpu_temp, 1) // ' C', observ=observ)
        rc = dm_observ_add_response(observ, name=NAME, unit='degC', value=cpu_temp)
    end subroutine add_cpu_temp

    subroutine add_disk_free(observ, path)
        !! Adds disk free and disk capacity responses to observation.
        character(*), parameter :: DISK_CAPACITY_NAME = 'disk_capacity'
        character(*), parameter :: DISK_FREE_NAME     = 'disk_free'

        type(observ_type), intent(inout) :: observ !! Observation.
        character(*),      intent(in)    :: path   !! File path of file system to inspect.

        character(256) :: file_system, message
        integer        :: rc, stat
        integer        :: capacity
        integer(i8)    :: available

        rc = dm_posix_disk_free(path, file_system=file_system, available=available, capacity=capacity)

        if (dm_is_error(rc)) then
            call logger%warning('failed to read free disk space', observ=observ, error=rc)
            return
        end if

        write (message, '("file system ", a, ": ", a, " available (", i0, " % used)")', iostat=stat) &
            trim(file_system), dm_size_to_human(available), capacity
        call logger%debug(message, observ=observ)

        rc = dm_observ_add_response(observ, name=DISK_CAPACITY_NAME, unit='%', value=capacity,  error=rc)
        rc = dm_observ_add_response(observ, name=DISK_FREE_NAME,     unit='B', value=available, error=rc)
    end subroutine add_disk_free

    subroutine add_load_average(observ, add_avg1, add_avg5, add_avg15)
        !! Add load average responses to observation.
        character(*), parameter :: AVG1_NAME  = 'load_avg1'
        character(*), parameter :: AVG5_NAME  = 'load_avg5'
        character(*), parameter :: AVG15_NAME = 'load_avg15'

        type(observ_type), intent(inout) :: observ    !! Observation.
        logical,           intent(in)    :: add_avg1  !! Add load average, 1 min.
        logical,           intent(in)    :: add_avg5  !! Add load average, 5 min.
        logical,           intent(in)    :: add_avg15 !! Add load average, 15 min.

        character(32) :: message
        integer       :: rc, stat
        real          :: avg1, avg5, avg15

        rc = dm_posix_load_average(avg1, avg5, avg15)

        if (dm_is_error(rc)) then
            call logger%warning('failed to read load average', observ=observ, error=rc)
            return
        end if

        write (message, '("load average:", 3(1x, f0.2))', iostat=stat) avg1, avg5, avg15
        call logger%debug(message, observ=observ)

        if (add_avg1)  rc = dm_observ_add_response(observ, name=AVG1_NAME,  unit='none', value=avg1,  error=rc)
        if (add_avg5)  rc = dm_observ_add_response(observ, name=AVG5_NAME,  unit='none', value=avg5,  error=rc)
        if (add_avg15) rc = dm_observ_add_response(observ, name=AVG15_NAME, unit='none', value=avg15, error=rc)
    end subroutine add_load_average

    subroutine add_log_db(observ, path)
        !! Adds log datatabase size response to observation.
        character(*), parameter :: NAME = 'log_db'

        type(observ_type), intent(inout) :: observ !! Observation.
        character(*),      intent(in)    :: path   !! Log database path.

        integer(i8) :: n

        n = dm_file_size(path, error=rc)

        if (dm_is_error(rc)) then
            call logger%warning('failed to read log database size', observ=observ, error=rc)
            return
        end if

        call logger%debug('log database ' // trim(path) // ': ' // dm_size_to_human(n), observ=observ)
        rc = dm_observ_add_response(observ, name=NAME, unit='B', value=n, error=rc)
    end subroutine add_log_db

    subroutine add_observ_db(observ, path)
        !! Adds observation datatabase size response to observation.
        character(*), parameter :: NAME = 'observ_db'

        type(observ_type), intent(inout) :: observ !! Observation.
        character(*),      intent(in)    :: path   !! Observation database path.

        integer(i8) :: n

        n = dm_file_size(path, error=rc)

        if (dm_is_error(rc)) then
            call logger%warning('failed to read observation database size', observ=observ, error=rc)
            return
        end if

        call logger%debug('observation database ' // trim(path) // ': ' // dm_size_to_human(n), observ=observ)
        rc = dm_observ_add_response(observ, name=NAME, unit='B', value=n, error=rc)
    end subroutine add_observ_db

    subroutine add_uptime(observ)
        !! Adds uptime response to observation.
        character(*), parameter :: NAME = 'uptime'

        type(observ_type), intent(inout) :: observ !! Observation.

        integer               :: rc
        integer(i8)           :: seconds
        type(time_delta_type) :: uptime

        call dm_posix_uptime(seconds, error=rc)

        if (dm_is_error(rc)) then
            call logger%warning('failed to read uptime', observ=observ, error=rc)
            return
        end if

        call dm_time_delta_from_seconds(uptime, seconds)
        call logger%debug('uptime: ' // dm_time_delta_to_string(uptime), observ=observ)
        rc = dm_observ_add_response(observ, name=NAME, unit='sec', value=seconds, error=rc)
    end subroutine add_uptime

    ! **************************************************************************
    ! COMMAND-LINE ARGUMENTS AND CONFIGURATION FILE.
    ! **************************************************************************
    integer function read_args(app) result(rc)
        !! Reads command-line arguments and settings from configuration file.
        type(app_type), intent(out) :: app !! App type.

        type(arg_parser_class) :: parser

        call parser%add('name',     short='n', type=ARG_TYPE_ID)      ! -n, --name <id>
        call parser%add('config',   short='c', type=ARG_TYPE_FILE)    ! -c, --config <path>
        call parser%add('logger',   short='l', type=ARG_TYPE_ID)      ! -l, --logger <id>
        call parser%add('node',     short='N', type=ARG_TYPE_ID)      ! -N, --node <id>
        call parser%add('sensor',   short='S', type=ARG_TYPE_ID)      ! -S, --sensor <id>
        call parser%add('target',   short='T', type=ARG_TYPE_ID)      ! -T, --target <id>
        call parser%add('receiver', short='r', type=ARG_TYPE_ID)      ! -r, --receiver <id>
        call parser%add('count',    short='C', type=ARG_TYPE_INTEGER) ! -C, --count <n>
        call parser%add('interval', short='I', type=ARG_TYPE_INTEGER) ! -I, --interval <n>
        call parser%add('debug',    short='D', type=ARG_TYPE_LOGICAL) ! -D, --debug
        call parser%add('verbose',  short='V', type=ARG_TYPE_LOGICAL) ! -V, --verbose

        ! Read all command-line arguments.
        rc = parser%read(version_callback)
        if (dm_is_error(rc)) return

        call parser%get('name',   app%name)
        call parser%get('config', app%config)

        ! Read configuration from file.
        rc = read_config(app)
        if (dm_is_error(rc)) return

        ! Get all other arguments.
        call parser%get('logger',   app%logger)
        call parser%get('node',     app%node_id)
        call parser%get('sensor',   app%sensor_id)
        call parser%get('target',   app%target_id)
        call parser%get('receiver', app%receiver)
        call parser%get('count',    app%count)
        call parser%get('interval', app%interval)
        call parser%get('debug',    app%debug)
        call parser%get('verbose',  app%verbose)

        ! Validate settings.
        rc = validate(app)
    end function read_args

    integer function read_config(app) result(rc)
        !! Reads configuration from file.
        type(app_type), intent(inout) :: app !! App type.

        type(config_class) :: config

        rc = E_NONE
        if (.not. dm_string_has(app%config)) return

        rc = config%open(app%config, app%name)

        if (dm_is_ok(rc)) then
            call config%get('logger',   app%logger)
            call config%get('node',     app%node_id)
            call config%get('sensor',   app%sensor_id)
            call config%get('target',   app%target_id)
            call config%get('receiver', app%receiver)
            call config%get('count',    app%count)
            call config%get('interval', app%interval)
            call config%get('debug',    app%debug)
            call config%get('verbose',  app%verbose)

            if (dm_is_ok(config%field('options'))) then
                call config%get('cpu_temp',   app%options%cpu_temp)
                call config%get('disk_free',  app%options%disk_free)
                call config%get('load_avg1',  app%options%load_avg1)
                call config%get('load_avg5',  app%options%load_avg5)
                call config%get('load_avg15', app%options%load_avg15)
                call config%get('log_db',     app%options%log_db)
                call config%get('observ_db',  app%options%observ_db)
                call config%get('uptime',     app%options%uptime)
                call config%remove()
            end if
        end if

        call config%close()
    end function read_config

    integer function validate(app) result(rc)
        !! Validates options and prints error messages.
        type(app_type), intent(inout) :: app !! App type.

        rc = E_INVALID

        if (dm_string_has(app%logger) .and. .not. dm_id_is_valid(app%logger)) then
            call dm_error_out(rc, 'invalid logger name')
            return
        end if

        if (.not. dm_id_is_valid(app%node_id)) then
            call dm_error_out(rc, 'invalid or missing node id')
            return
        end if

        if (.not. dm_id_is_valid(app%sensor_id)) then
            call dm_error_out(rc, 'invalid or missing sensor id')
            return
        end if

        if (.not. dm_id_is_valid(app%target_id)) then
            call dm_error_out(rc, 'invalid or missing target id')
            return
        end if

        if (dm_string_has(app%receiver) .and. .not. dm_id_is_valid(app%receiver)) then
            call dm_error_out(rc, 'invalid receiver')
            return
        end if

        if (app%count < 0) then
            call dm_error_out(rc, 'invalid count')
            return
        end if

        if (app%interval <= 0) then
            call dm_error_out(rc, 'invalid interval')
            return
        end if

        rc = E_NOT_FOUND

        if (dm_string_has(app%options%disk_free) .and. .not. dm_file_exists(app%options%disk_free)) then
            call dm_error_out(rc, 'disk free path does not exist')
            return
        end if

        if (dm_string_has(app%options%log_db) .and. .not. dm_file_exists(app%options%log_db)) then
            call dm_error_out(rc, 'log database does not exist')
            return
        end if

        if (dm_string_has(app%options%observ_db) .and. .not. dm_file_exists(app%options%observ_db)) then
            call dm_error_out(rc, 'observation database does not exist')
            return
        end if

        rc = E_NONE
    end function validate

    ! **************************************************************************
    ! CALLBACKS.
    ! **************************************************************************
    subroutine signal_callback(signum) bind(c)
        !! C-interoperable signal handler that stops the program.
        integer(c_int), intent(in), value :: signum !! Signal number.

        call logger%debug('exit on on signal ' // dm_posix_signal_name(signum))
        call shutdown(E_NONE)
    end subroutine signal_callback

    subroutine version_callback()
        call dm_version_out(APP_NAME, APP_MAJOR, APP_MINOR, APP_PATCH)
        print '(a)', dm_lua_version(.true.)
    end subroutine version_callback
end program dmsystem
